### ISSUE TITLE
Destructive substitution on classes produces unrepresentable module type

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6654]
**Original bug ID:** 6654
**Reporter:** @lpw25
**Status:** confirmed (set by @damiendoligez on 2015-01-07T18:57:14Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Target version:** later
**Category:** typing
**Related to:** #4791
**Monitored by:** @hcarty

## Bug description

If you apply destructive substitution to a class or class type the type system handles it correctly, but the resulting module type cannot be represented by OCaml's syntax:

    # module type S = sig
        class c : object method
        m : int end
      end;;
    module type S = sig class c : object method m : int end end

    # class d = object method m = 3 end;;
    class d : object method m : int end

    # module type T = S with type c := d;;
    module type T =
      sig
        class c : object method m : int end
        class type c = object method m : int end
        type #c = #d
      end

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Yet another example of signature that cannot be parsed/typed.
Of course we could forbid it, but would it help?
Another solution would be to allow destructive substitution for all declaration, but this is harder.
For the time being, doing nothing seems reasonable.

**Comment by yallop** [comment]
This issue has turned into a bug in `trunk`.  The final line now fails with an internal `Invalid_argument` exception:

```
# module type S = sig
    class c : object method
    m : int end
  end;;
      module type S = sig class c : object method m : int end end
# class d = object method m = 3 end;;
class d : object method m : int end
# module type T = S with type c := d;;
Fatal error: exception Invalid_argument("split_at")
```

**Comment by gasche** [comment]
This suggests that maybe we should be more proactive in adding examples for known issues in the testsuite, so that we notice when the behavior changes (in good or bad ways).

**Comment by garrigue** [comment]
There is a real bug now. Appeared in 4.08.

**Comment by Octachron** [comment]
Fixed in 4.13 by #10401 : `with` constraint cannot alter the type associated to a class anymore, and

```ocaml
# module type T = S with type c := d;;
```
now fails with

> Error: The signature constrained by `with' has no component named c
